### DISCUSSION TITLE
Integrate tests into a “make test” target

### DISCUSSION THREAD

**Discussion by comburent** [jcorporation/myMPD#1022] [Ideas]
Hi,

Would you consider using CMake's CTest module in order to have the standard `make test` target from the default CMakeLists.txt?
These automated tests are useful for distributions where package testing is considered a part of the build process and helps spotting issues before the updates are sent.

**Comment by jcorporation** [comment]
You mean the unit tests?

**Reply by comburent** [reply]
Yes

**Reply by jcorporation** [reply]
Implemented in the v10.4.0 branch, Set `MYMPD_BUILD_TESTING=ON` and there will be a `make test` target.

**Reply by comburent** [reply]
Could the tests themselves get labels/properties individually rather than be lumped together into a single `UnitTest` test? <https://cmake.org/cmake/help/book/mastering-cmake/chapter/Testing%20With%20CMake%20and%20CTest.html#how-does-cmake-facilitate-testing>
Running `ctest -N` gives:
```
  Test #1: UnitTest

Total Tests: 1 
```

Having the _unit tests_ under tests/ as individual cmake _tests_  makes it easier to exclude certain tests with `ctest --exclude-regex ...` . (or individually run each test during development)
From a distribution perspective, the build phase of a package is usually run without internet access so tests like `http_client.test_http_client` would fail unless excluded.

**Reply by jcorporation** [reply]
This involves a complete redesign of the test suite.

**Reply by comburent** [reply]
> This involves a complete redesign of the test suite.

Would you consider it?
The test can be patched from the distribution side though it's best if things can be solved directly upstream.

**Reply by jcorporation** [reply]
Yes, I would accept patches.

**Reply by jcorporation** [reply]
The unit test framework seems to support running specific tests. We could simply expose this.

https://github.com/sheredom/utest.h

Output after this small patch (6316c3e07653dc0184563bd777007bbd091ce08b):
```
ctest -N
Test project /tmp/myMPD/debug
  Test  #1: test_api
  Test  #2: test_env
  Test  #3: test_filehandler
  Test  #4: test_http_client
  Test  #5: test_jsonrpc
  Test  #6: test_list
  Test  #7: test_m3u
  Test  #8: test_mimetype
  Test  #9: test_mpd_client_tags
  Test #10: test_mympd_queue
  Test #11: test_mympd_state
  Test #12: test_radix_sort
  Test #13: test_random
  Test #14: test_sds_extras
  Test #15: test_state_files
  Test #16: test_timer
  Test #17: test_utility
  Test #18: test_validate
  Test #19: test_cert
  Test #20: test_lyrics_id3
  Test #21: test_lyrics_flac

Total Tests: 21
```

**Reply by comburent** [reply]
Nice!
